### DISCUSSION TITLE
Stage failure on an actor

### DISCUSSION THREAD

**Discussion by valentina-2000** [akkadotnet/akka.net#7339] [Q&A]
Hi, I have a question. If I've got a graph stage running inside an actor and the stage fails, is the actor aware abotut the failed stage?
How can I catch this exception inside the actor?

**Comment by richardjharding** [comment]
@valentina-2000 I responded to this on the discord before I realised you raised it here - I have used the PipeTo extension on the task to have any messages including extensions sent to the "Self" actor ref and handle them there - I log them and attempt to restart the flow

I think you can also do this with stream Attributes but I've not found a good example yet

**Comment by Aaronontheweb** [comment]
@richardjharding 's answer is good, but another tool you can use is the [`WatchTermination` stream stage](https://getakka.net/articles/streams/builtinstages.html#watchtermination). This stage exposes a `Task` that will complete when the stream stage terminates in both failure and successful cases.

In a failure case that `Task` will fail with an `Exception` - you can pipe that failure back to the parent actor.

https://github.com/akkadotnet/akka.net/blob/8e2e11d7899acc13a96e9c6821c66455a8915ed0/src/core/Akka.Streams.Tests/Dsl/FlowWatchTerminationSpec.cs#L86-L97

` var future = t.Item1;` <- this is the termination `Task`
